app/presenters/grouped_result_set_presenter.rb

```diff
@@ -27,45 +24,69 @@ def grouped_documents
     # duplication in every primary facet value grouping.
     default_documents, other_documents = sorted_documents.partition do |document|
-      filtered_by_primary_facet? && tagged_to_all?(primary_facet_key, document)
+      selected_values_in_primary_facet.present? && tagged_to_all?(primary_facet_key, document)
     end
 
     default_group = default_documents.empty? ? [] : [{ facet_key: "all_businesses", facet_name: "All businesses", documents: default_documents }]
 
-    other_documents.each do |item|
-      document_metadata = item[:document][:metadata]
-      document_metadata.each do |metadata|
-        key = metadata[:id]
-        next unless key && facet_filters.has_key?(key.to_sym)
-
-        if primary_facet_key == key
-          # Group by value for the primary facet
-          (metadata[:labels] & facet_filters[primary_facet_key.to_sym]).each do |value|
-            populate_group(primary_group, value, facet_label_for(value), item)
-          end
-        else
-          # Group by facet name otherwise
-          populate_group(secondary_group, key, metadata[:label], item)
-        end
-      end
+    unsorted_primary_group = selected_values_in_primary_facet.map do |selected_value|
+      documents = documents_tagged_to_primary_facet_value(other_documents, selected_value)
+      {
+        facet_key: selected_value,
+        facet_name: label_for_facet_value(selected_value),
+        documents: documents
+      }
+    end
+
+    primary_group = unsorted_primary_group.sort_by { |group| group[:group_name] }
+
+    secondary_group = secondary_facets.map do |secondary_facet|
+      documents = documents_tagged_to_secondary_facet(other_documents, secondary_facet.key)
+      {
+        facet_key: secondary_facet.key,
+        facet_name: label_from_metadata(documents.first, secondary_facet.key),
+        documents: documents
+      }
     end
 
-    compact_and_sort(primary_group) + compact_and_sort(secondary_group, true) + default_group
+    results = primary_group + secondary_group + default_group
+    results.reject { |result| result[:documents].empty? }
   end
 
 private
 
+  def label_from_metadata(document, key)
+    return if document.nil?
+
+    metadata = document[:document][:metadata].find { |m| m[:id] == key }
+    metadata[:label]
+  end
+
+  def documents_tagged_to_primary_facet_value(documents, selected_value)
+    documents.select do |document|
+      document[:document][:metadata].any? do |metadata|
+        metadata[:id] == primary_facet_key &&
+          metadata[:labels].include?(selected_value)
+      end
+    end
+  end
+
+  def documents_tagged_to_secondary_facet(documents, secondary_group_name)
+    documents.select do |document|
+      document[:document][:metadata].any? { |metadata| metadata[:id] == secondary_group_name }
+    end
+  end
+
   def primary_facet_key
     finder.facets.first.key
   end
 
-  def filtered_by_primary_facet?
-    facet_filters.key?(primary_facet_key.to_sym)
+  def primary_facet
```

Comment:
Is `primary_facet` used anywhere?

Reply:
oh, you're right!! Another 3 lines removed! 👍